pkg/tracer/tracer_linux.go

```diff
@@ -479,6 +567,86 @@ func (t *LinuxTracer) getMap(mapName string) (*bpflib.Map, error) {
 	return mp, nil
 }
 
+func (t *LinuxTracer) dispatchPerfEvent(event *common.PerfEvent) {
+	t.tcpConnInsightsLock.Lock()
+	defer t.tcpConnInsightsLock.Unlock()
+
+	if event.HTTPResponse != nil {
+		logger.Tracef("http response: %v", event.HTTPResponse)
+		httpRes := event.HTTPResponse
+
+		conn, ok := t.tcpConnInsights[httpRes.Connection]
+		httpProtocol := "http"
+		if !ok {
+			conn = ConnInsight{
+				ApplicationProtocol: httpProtocol,
+				HttpMetrics:         make(map[HttpCode]*ddsketch.DDSketch),
+			}
+		}
+		conn.ApplicationProtocol = httpProtocol
+
+		latencyCounter, ok := conn.HttpMetrics[httpRes.StatusCode]
+		if !ok {
+			var err error
+			latencyCounter, err = ddsketch.NewDefaultDDSketch(t.config.HttpMetricPrecision)
+			if err != nil {
+				logger.Errorf("can't create dd sketch")
+			}
+			conn.HttpMetrics[httpRes.StatusCode] = latencyCounter
+		}
+		err := latencyCounter.Add(httpRes.ResponseTime.Seconds())
+		if err != nil {
+			logger.Errorf("can't count")
+		}
+		t.tcpConnInsights[httpRes.Connection] = conn
+
+	} else if event.MySQLGreeting != nil {
+		logger.Tracef("mysql greeting: %v", event.MySQLGreeting)
+		mysqlGreeting := event.MySQLGreeting
+
+		conn, ok := t.tcpConnInsights[mysqlGreeting.Connection]
+		if !ok {
+			conn = ConnInsight{
+				ApplicationProtocol: "mysql",
+				HttpMetrics:         make(map[HttpCode]*ddsketch.DDSketch),
+			}
+		}
+		conn.ApplicationProtocol = "mysql"
+		t.tcpConnInsights[mysqlGreeting.Connection] = conn
+	}
+}
+
+func (t *LinuxTracer) enrichTcpConns(conns []common.ConnectionStats) []common.ConnectionStats {
+	logger.Debug("enrich tcp connections")
+	for i := range conns {
+		conn := conns[i]
+		t.tcpConnInsightsLock.Lock()
```

Comment:
Extract the block that is locked to a function and do the following instead of unlocking at the end of the block. 

```golang
t.tcpConnInsightsLock.Lock()
defer t.tcpConnInsightsLock.Unlock()
```

So that you're sure that even in unexpected scenarios the lock is still released.

Reply:
Done